src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -563,7 +563,8 @@ private static async void ParallelConnectHelper(
             {
                 // Try to connect.  If we're successful, store this task into the result task.
                 await connectTask.ConfigureAwait(false);
-                success = tcs.TrySetResult(socket);
+                // A connection attempt could return quickly if a server refuses connection or there is some other error
+                success = tcs.TrySetResult(socket) && socket.Connected;
```

Comment:
```suggestion
                success = socket.Connected && tcs.TrySetResult(socket);
```

We'll want to make sure that the socket is connected before setting the TCS result. We can only set the TCS result once, so if we set it before checking if the socket is connected then we might set it to a disconnected socket. In that situation, `success` would be false for every socket, even the ones which successfully connected - so connection establishment would fail.